File: tests/test_base91x.py
"""
MIT License

Copyright (c) 2026 Roman Babenko

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""
import binascii
import json
import random
import string
from pathlib import Path
from random import randbytes

from base91x import base91x

TEXT = "The quick brown\r\nfox\tjumps\nover\rthe lazy\n\rdog!"
PANGRAM = "Thequickbrownfoxjumpsoverthelazydog!"
DATA = [88, 136, 162, 112, 31, 156, 195, 75, 208, 5, 61, 106, 20, 163, 227, 172, 240, 150, 163, 100, 63, 170, 82,
        175, 58, 17, 203, 5, 3]


def test_static_alphabet():
    assert len(base91x.BASE91X_ALPHABET) == base91x.BASE91X_LEN


def test_static_decode():
    data = base91x.decode(TEXT)
    assert list(data) == DATA


def test_static_encode():
    text = base91x.encode(bytes(DATA))
    assert text == PANGRAM


def test_refurbish_small():
    for n in range(33):
        original_data = randbytes(n)
        text = base91x.encode(original_data)
        refurbish_data = base91x.decode(text)
        assert len(original_data) == len(refurbish_data)
        assert original_data == refurbish_data, text


def test_refurbish_large():
    original_data = randbytes(65536)
    text = base91x.encode(original_data)
    refurbish_data = base91x.decode(text)
    assert len(original_data) == len(refurbish_data)
    assert original_data == refurbish_data, text


def test_stress_full_decode():
    text = ""
    text_size = random.randint(0, 65536)
    while text_size > len(text):
        text += chr(random.randint(0, 0x10FFFF))
    assert len(text) == text_size
    data = base91x.decode(text)
    assert len(data) <= text_size


def test_stress_ascii_decode():
    text = ""
    text_size = random.randint(0, 65536)
    while text_size > len(text):
        text += random.choice(string.printable)
    assert len(text) == text_size
    data = base91x.decode(text)
    assert len(data) <= text_size


def test_constants():
    # ## preparation
    # test_dict = {}
    # for n in range(1, 64):
    #     for m in range(32):
    #         original_data = randbytes(n)
    #         test_dict[binascii.hexlify(original_data).decode()] = base91x.encode(original_data)
    # with open(Path(__file__).parent / "test.json", 'w') as o:
    #     json.dump(test_dict, o, indent=4, sort_keys=True)
    # del test_dict
    # ## check algorithm stability
    with open(Path(__file__).parent / "test.json") as f:
        test_dict = json.load(f)
    for hex_data, base91x_data in test_dict.items():
        data = binascii.unhexlify(hex_data)
        new_base91x = base91x.encode(data)
        assert new_base91x == base91x_data


def test_endings():
    # zero produces zero
    assert base91x.encode(b'') == ""
    assert base91x.encode(b'A') == ">!"
    assert base91x.encode(b'AA') == "O|}"
    assert base91x.encode(b'AAA') == "O|<z"
    assert base91x.encode(b'AAAA') == "O|ico"

    # huge empty tail for 5 bytes input
    assert base91x.encode(b'AAAAA') == "O|ic.R!"
    # various ends may impact on result
    assert base91x.decode("O|ic.Ro") == b'AAAAA'
    assert base91x.decode("O|ic.Rp") == b'AAAA\xc1'
    assert base91x.decode("O|ic.Rr") == b'AAAA\xc1'
    assert base91x.decode("O|ic.Rs") == b'AAAAA'
    assert base91x.decode("O|ic.Rt") == b'AAAA\xc1'

    assert base91x.encode(b'AAAAAA') == "O|ic.RX~"
    assert base91x.encode(b'AAAAAAA') == "O|ic.Rzx{"
    assert base91x.encode(b'AAAAAAAA') == "O|ic.RzxTt"
    assert base91x.encode(b'AAAAAAAAA') == "O|ic.RzxSG_!"
    assert base91x.encode(b'AAAAAAAAAA') == "O|ic.RzxSG:~~"
    assert base91x.encode(b'AAAAAAAAAAA') == "O|ic.RzxSG:~0}"
    assert base91x.encode(b'AAAAAAAAAAAA') == "O|ic.RzxSG:~tqw"

    # 13 bytes are encoded exactly to 16 symbols (8 words)
    assert base91x.encode(b'AAAAAAAAAAAAA') == "O|ic.RzxSG:~tq)i"
    # extra ending has no impact
    assert base91x.decode("O|ic.RzxSG:~tq)i~") == b'AAAAAAAAAAAAA'

File: tests/test_base91x.cpp
/*
MIT License

Copyright (c) 2023 Roman Babenko

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
*/
#include <iostream>
#include <algorithm>
#include <cstdlib>
#include <ctime>
#include <chrono>
#include <cmath>

#include "base91x.hpp"

#define LOG std::cerr << "L_" << __LINE__ << ": "

//------------------------------------------------------------------------------

bool in_abc(char i)
{
    return ((92 < i and i < 127) or (39 < i and i < 92) or (36 < i and i < 39) or 33 == i or 35 == i or 36 == i);
}

//------------------------------------------------------------------------------

bool test_alphabets()
{
    if (91 != sizeof(base91x::BASE91X_ALPHABET))
    {
        LOG << "Wrong size of direct alphabet "<<  sizeof(base91x::BASE91X_ALPHABET) << std::endl;
        return false;
    }

    if (0x80 != sizeof(base91x::BASE91X_ZYX))
    {
        LOG << "Wrong size of reverse alphabet" << std::endl;
        return false;
    }

    for (unsigned n = 0; n < sizeof(base91x::BASE91X_ZYX); ++n)
    {
        char digit = -2;
        if ((92 < n and n < 127) or (39 < n and n < 92) or (36 < n and n < 39))
            digit = 0x7F ^ n;
        else if (33 == n)
            digit = 0;
        else if (35 == n)
            digit = 35;
        else if (36 == n)
            digit = 88;
        else
            digit = -1;

        if (digit != base91x::BASE91X_ZYX[n])
        {
            LOG << "error on " << n << " element" << std::endl;
            return false;
        }
    }
    for (auto &i : base91x::BASE91X_ALPHABET)
    {
        if (sizeof(base91x::BASE91X_ZYX) <= i)
        {
            LOG << "out of size " << (i) << " element" << std::endl;
            return false;
        }
        if (0 > base91x::BASE91X_ZYX[i])
        {
            LOG << "error on " << i << " element" << std::endl;
            return false;
        }
        if (not in_abc(i))
        {
            LOG << "wrong " << i << " element" << std::endl;
            return false;
        }
    }

    //        const char digit = BASE91X_ZYX[0x7F & i];

    LOG << "Alphabets tests passed" << std::endl;
    return true;
}

//------------------------------------------------------------------------------

template <typename T>
bool test_refurbish(const size_t size)
{
    T data_original(size);
    std::generate(data_original.begin(), data_original.end(), std::rand);

    std::string text;
    base91x::encode(data_original, text);

    size_t n = 0;
    // verify whether encoded text contains symbols from base91x alphabet only
    for (auto it_text : text)
    {
        if (not in_abc(it_text))
        {
            LOG << "error on " << n << " element '" << it_text << "' not in base91x::BASE91X_ALPHABET" << std::endl;
            return false;
        }
        ++n;
    }

    LOG << "Test with data [" << data_original.size() << "] -> encoded text [" << text.size() << "]" << std::endl;
    if (80 < text.size())
    {
        LOG << "Limit to first 80 symbols:" << std::endl << text.substr(1, 79) << std::endl;
    }
    else
    {
        LOG << text << std::endl;
    }

    T data_refurbed;
    base91x::decode(text, data_refurbed);

    if (data_original.size() != data_refurbed.size())
    {
        LOG << "data_original size (" << data_original.size() << ") does not equal data_refurbed size ("
            << data_refurbed.size() << ")" << std::endl;
        return false;
    }

    n = 0;
    auto it_original = data_original.begin();
    auto it_refurbed = data_refurbed.begin();
    while (it_original != data_original.end() and it_refurbed != data_refurbed.end())
    {
        if (*it_original != *it_refurbed)
        {
            LOG << "error on " << n << " element" << std::endl;
            return false;
        }
        ++n;
        ++it_original;
        ++it_refurbed;
    }
    LOG << "Test passed with size = " << data_original.size() << std::endl;

    if (base91x::compute_encoded_size(size) != text.size())
    {
        LOG << "computed " << base91x::compute_encoded_size(size) << " != " << text.size() << std::endl;
        return false;
    }

    if (base91x::assume_decoded_size(text.size()) != size)
    {
        LOG << "assumed " << base91x::assume_decoded_size(size) << " != " << text.size() << std::endl;
        return false;
    }

    return true;
}

//------------------------------------------------------------------------------

bool test_static_with_space()
{
    const std::string_view pangram_test = "* / * / *The quick brown fox\tjumps\nover\rthe lazy dog { < # > }";
    const std::string pangram_expected = "*/*/*Thequickbrownfoxjumpsoverthelazydog{<#>}";
    std::vector<unsigned char> data;
    base91x::decode(pangram_test, data);

    const std::vector<unsigned char> data_expected = {197, 188, 152, 123, 190, 170, 196, 57, 20, 212, 152, 234, 45,
        19, 38, 185, 248, 29, 56, 51, 69, 134, 70, 46, 193, 65, 219, 166, 60, 124, 38, 76, 84, 125, 125, 174};

    if (data_expected.size() != data.size())
    {
        LOG << "data size = " << data.size() << ": expected size = " << data_expected.size() << std::endl;
        return false;
    }

    size_t n = 0;
    auto it = data.begin();
    auto it_ = data_expected.begin();
    while (it != data.end() and it_ != data_expected.end())
    {
        if (*it != *it_)
        {
            LOG << "Error on " << n << " element. got " << int(*it_) << " expected " << int(*it) << std::endl;
            return false;
        }
        ++n;
        ++it;
        ++it_;
    }

    if (data.size() != n)
    {
        LOG << "Pass only " << n << "bytes. Expected " << data.size() << std::endl;
        return false;
    }

    std::string pangram_refurb;
    base91x::encode(data, pangram_refurb);

    if (pangram_expected != pangram_refurb)
    {
        LOG << "Not matched" << std::endl << pangram_expected << std::endl << pangram_refurb << std::endl;
        return false;
    }
    LOG << "Static test passed" << std::endl;
    return true;
}

//------------------------------------------------------------------------------

bool test_stress(const size_t size)
{
    std::string text;
    text.resize(size);
    std::generate(text.begin(), text.end(), std::rand);
    std::vector<unsigned char> data;
    base91x::decode(text, data);
    LOG << "Stress test: from " << size << " symbols were decoded " << data.size() << " bytes" << std::endl;
    LOG << "Stress test passed" << std::endl;
    return true;
}

//------------------------------------------------------------------------------

int main(const int argc, const char *argv[])
{
    std::srand(std::time(nullptr));
    // various size of random data to cover all possibly calculated size
    for (int n = 0; n < 33; ++n)
    {
        if (not( //
                test_refurbish<std::vector<char>>(n) //
                && //
                test_refurbish<std::vector<unsigned char>>(n) //
                ))
            return 1;
    }
    // 1Mb of random data
    if ( //
        test_refurbish<std::vector<char>>(std::rand() % (1 << 20)) //
        && //
        test_refurbish<std::vector<unsigned char>>(std::rand() % (1 << 20)) //
        && //
        test_static_with_space() //
        && //
        test_alphabets() //
        && //
        test_stress(1 << 20) //
    )
    {
        return 0;
    }
    return 1;
}

//------------------------------------------------------------------------------
